File: dsp_2026_2/19_modulacion_am.m
% 19_modulacion_am.m
% Tema 3 - Modulacion y demodulacion AM (desplazamiento del espectro).
% Ejecutar:  octave-cli 19_modulacion_am.m
%
% Referencias a las laminas del Tema 3:
%   - El espectro de la senal AM muestra portadora (fc) y bandas laterales
%     (fc +/- fm).
%   - Demodulacion simple: multiplicar por la portadora desplaza el espectro
%     a banda base y a 2fc; se recupera con un filtro ideal en frecuencia.
%
% Senal AM:  x(t) = (1 + mu*cos(2*pi*fm*t))*cos(2*pi*fc*t).

clear; close all; clc;

fc = 100e3;   % portadora [Hz]
fm = 1e3;     % modulante [Hz]
fs = 400e3;   % muestreo [Hz]
N = 4000;     % N tal que fc y fm caen en bins exactos
mu = 0.5;
t = (0:N-1)/fs;

% ------------------- modulacion -------------------
m = cos(2*pi*fm*t);                    % senal modulante
x = (1 + mu*m) .* cos(2*pi*fc*t);      % senal AM

X = fft(x);
f = (0:N-1)*fs/N;
Xm = abs(X)*2/N;

fprintf('============================================================\n');
fprintf('MODULACION AM\n');
fprintf('============================================================\n');
fprintf('fc = %.0f Hz, fm = %.0f Hz, mu = %.1f\n', fc, fm, mu);
fprintf('espectro: picos en fc, fc-fm y fc+fm (%.0f, %.0f, %.0f Hz)\n', ...
        fc, fc-fm, fc+fm);

% ------------------- demodulacion (simple) -------------------
xr = x .* cos(2*pi*fc*t);      % mezcla con la portadora
Xr = fft(xr);
Xrm = abs(Xr)*2/N;             % espectro desplazado

fprintf('\nDEMODULACION (simple): ESPECTRO DESPLAZADO\n');
fprintf('============================================================\n');
fprintf('tras mezclar con la portadora, el espectro se desplaza a banda base\n');
fprintf('(picos en 0 y +/-fm) y a 2fc (2fc +/- fm).\n');

% ---------- Una sola figura con los 3 plots ----------
figure;

% 1) AM en tiempo (primeros 3 ms)
subplot(3,1,1);
idx = t <= 3e-3;
plot(t(idx)*1e3, x(idx));
title('Senal AM (temporal)'); xlabel('t [ms]'); grid on;

% 2) AM en frecuencia [0, fs/2]
subplot(3,1,2);
plot(f(1:N/2+1)/1e3, Xm(1:N/2+1));
title('Espectro AM (portadora + bandas laterales)'); xlabel('f [kHz]'); grid on;
hold on;
plot([fc fc]/1e3, ylim, 'r--');
plot([(fc-fm) (fc-fm)]/1e3, ylim, 'g:');
plot([(fc+fm) (fc+fm)]/1e3, ylim, 'g:');
legend('espectro', 'fc', 'fc±fm');

% 3) Demodulacion en frecuencia [0, fs), incluye 2fc
subplot(3,1,3);
plot(f/1e3, Xrm);
title('Espectro tras mezclar: baseband, 2fc y espejo'); xlabel('f [kHz]'); grid on;
hold on;
plot([0 0], ylim, 'b--');
plot([fm fm]/1e3, ylim, 'b:');
plot([2*fc 2*fc]/1e3, ylim, 'r--');
plot([(2*fc-fm) (2*fc-fm)]/1e3, ylim, 'g:');
plot([(2*fc+fm) (2*fc+fm)]/1e3, ylim, 'g:');
legend('espectro', '0', 'fm', '2fc', '2fc±fm');

File: dsp_2026_2/19_modulacion_am.py
#!/usr/bin/env python3
"""19_modulacion_am.py

Tema 3 - Modulacion y demodulacion AM.
Ejecutar:  python3 19_modulacion_am.py

Referencias a las laminas del Tema 3:
  - El espectro de la senal AM muestra portadora (fc) y bandas laterales
    (fc +/- fm).
  - Demodulacion coherente: multiplicar por la portadora y filtrar el paso
    bajo para recuperar la senal modulante.

Senal AM:  x(t) = (1 + mu*cos(2*pi*fm*t))*cos(2*pi*fc*t).
"""

import matplotlib.pyplot as plt
import numpy as np

fc = 100e3     # portadora [Hz]
fm = 1e3       # modulante [Hz]
fs = 400e3     # frecuencia de muestreo [Hz]
N = 4000       # N tal que fc y fm caen en bins exactos
mu = 0.5       # indice de modulacion
t = np.arange(N)/fs

# ------------------- modulacion -------------------
m = np.cos(2*np.pi*fm*t)                     # senal modulante
x = (1 + mu*m) * np.cos(2*np.pi*fc*t)        # senal AM

X = np.fft.fft(x)
f = np.arange(N)*fs/N
Xm = np.abs(X)*2/N

print("="*58)
print("MODULACION AM")
print("="*58)
print("fc = %.0f Hz, fm = %.0f Hz, mu = %.1f" % (fc, fm, mu))
print("espectro: picos en fc, fc-fm y fc+fm (%.0f, %.0f, %.0f Hz)"
      % (fc, fc-fm, fc+fm))

# ------------------- demodulacion (simple) -------------------
xr = x * np.cos(2*np.pi*fc*t)          # mezcla con la portadora
Xr = np.fft.fft(xr)
Xrm = np.abs(Xr)*2/N                   # espectro desplazado

print("\nDEMODULACION (simple): ESPECTRO DESPLAZADO")
print("="*58)
print("tras mezclar con la portadora, el espectro se desplaza a banda base")
print("(picos en 0 y +/-fm) y a 2fc (2fc +/- fm).")

# ---------- Una sola figura con los 3 plots ----------
plt.figure(figsize=(10, 10))

# 1) AM en tiempo (primeros 3 ms)
plt.subplot(3, 1, 1)
idx = t <= 3e-3
plt.plot(t[idx]*1e3, x[idx])
plt.title('Señal AM (temporal)'); plt.xlabel('t [ms]'); plt.grid()

# 2) AM en frecuencia [0, fs/2]
plt.subplot(3, 1, 2)
plt.plot(f[:N//2 + 1]/1e3, Xm[:N//2 + 1])
plt.title('Espectro AM (portadora + bandas laterales)')
plt.xlabel('f [kHz]'); plt.grid()
plt.axvline(fc/1e3, color='r', ls='--', alpha=0.5, label='fc')
plt.axvline((fc-fm)/1e3, color='g', ls=':', alpha=0.5, label='fc±fm')
plt.axvline((fc+fm)/1e3, color='g', ls=':', alpha=0.5)
plt.legend()

# 3) Demodulacion en frecuencia [0, fs), incluye 2fc
plt.subplot(3, 1, 3)
plt.plot(f/1e3, Xrm)
plt.title('Espectro tras mezclar: baseband, 2fc y espejo')
plt.xlabel('f [kHz]'); plt.grid()
plt.axvline(0, color='b', ls='--', alpha=0.3)
plt.axvline(fm/1e3, color='b', ls=':', alpha=0.3)
plt.axvline(2*fc/1e3, color='r', ls='--', alpha=0.5, label='2fc')
plt.axvline((2*fc-fm)/1e3, color='g', ls=':', alpha=0.5, label='2fc±fm')
plt.axvline((2*fc+fm)/1e3, color='g', ls=':', alpha=0.5)
plt.axvline(fs/1e3, color='k', ls='-', alpha=0.2)
plt.text(fs/2/1e3, plt.ylim()[1]*0.9, 'fs/2', ha='center', fontsize=8, color='k', alpha=0.4)
plt.legend()

plt.tight_layout(); plt.show()
